dask/array/linalg.py

```diff
@@ -327,9 +421,23 @@ def qr(a, name=None):
     --------
 
     np.linalg.qr : Equivalent NumPy Operation
-    dask.array.linalg.tsqr: Actual implementation with citation
+    dask.array.linalg.tsqr: Implementation for tall-and-skinny arrays
+    dask.array.linalg.sfqr: Implementation for short-and-fat arrays
     """
-    return tsqr(a, name)
+
+    if len(a.chunks[1]) == 1:
+        return tsqr(a, name)
+    elif len(a.chunks[0]) == 1:
+        return sfqr(a, name)
+    else:
+        raise NotImplemented(
+            "qr currently supports only tall-and-skinny (single column chunk/block; see tsqr)\n"
+            "and short-and-fat (single row chunk/block; see sfqr) matrices\n\n"
+            "Consider use of the rechunk method. For example,\n\n"
+            "x.rechunk({0: -1, 1: 'auto'}) or x.rechunk({0: 'auto', 1: -1})\n\n"
+            "which rechunk one shorter axis to a single chunk, while allowing\n"
+            "the other axis to automatically grow/shrink appropriately."
+        )
```

Comment:
I suspect that you wanted NotImplementedError here.  

```python
In [1]: import dask.array as da

In [2]: x = da.ones((100, 100), chunks=(10, 10))

In [3]: da.linalg.qr(x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-a349040cfc1e> in <module>()
----> 1 da.linalg.qr(x)

~/workspace/dask/dask/array/linalg.py in qr(a, name)
    432     else:
    433         raise NotImplemented(
--> 434             "qr currently supports only tall-and-skinny (single column chunk/block; see tsqr)\n"
    435             "and short-and-fat (single row chunk/block; see sfqr) matrices\n\n"
    436             "Consider use of the rechunk method. For example,\n\n"

TypeError: 'NotImplementedType' object is not callable
```

We should probably add a test for this to makes sure that this works properly.

I think that we should keep the function names `tsqr` and `sfqr` hidden from top-level users if possible.  Even though they're pretty clear to those who have some experience with blocked linear algebra they might not be broadly understood among all users.

Reply:
Should be a `NotImplementedError`... I'll edit that.